cluster/src/main/java/org/apache/iotdb/cluster/log/manage/serializable/SyncLogDequeSerializer.java

```diff
@@ -0,0 +1,310 @@
+   */
+  public SyncLogDequeSerializer(String logPath) {
+    logFile = SystemFileFactory.INSTANCE
+        .getFile(logPath + File.separator + "logData");
+    metaFile = SystemFileFactory.INSTANCE
+        .getFile(logPath + File.separator + "logMeta");
+    String name = logFile.getAbsolutePath();
+    init();
+  }
+
+  /**
+   * log in disk is
+   * size of log | log buffer
+   * meta in disk is
+   * firstLogPosition | size of log meta | log meta buffer
+   */
+  public SyncLogDequeSerializer() {
+    String systemDir = IoTDBDescriptor.getInstance().getConfig().getSystemDir();
+    logFile = SystemFileFactory.INSTANCE
+        .getFile(systemDir + File.separator + "raftLog" + File.separator + "logData");
+    metaFile = SystemFileFactory.INSTANCE
+        .getFile(systemDir + File.separator + "raftLog" + File.separator + "logMeta");
+    init();
+  }
+
+  // only for test
+  public void setMaxRemovedLogSize(long maxRemovedLogSize) {
+    this.maxRemovedLogSize = maxRemovedLogSize;
+  }
+
+  public Deque<Integer> getLogSizeDeque() {
+    return logSizeDeque;
+  }
+
+  // init output stream
+  private void init() {
+    try {
+      if (!logFile.getParentFile().exists()) {
+        logFile.getParentFile().mkdir();
+        logFile.createNewFile();
+        logFile.createNewFile();
+      }
+
+      logOutputStream = new FileOutputStream(logFile, true);
+      metaOutputStream = new FileOutputStream(metaFile, true);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public void addLast(Log log, LogManagerMeta meta) {
+    ByteBuffer data = log.serialize();
+    int totalSize = 0;
+    // write into disk
+    try {
+      totalSize += ReadWriteIOUtils.write(data, logOutputStream);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+    logSizeDeque.addLast(totalSize);
+    serializeMeta(meta);
+  }
+
+  @Override
+  public void removeLast(LogManagerMeta meta) {
+    truncateLogIntern(1);
+    serializeMeta(meta);
+  }
+
+  @Override
+  public void truncateLog(int count, LogManagerMeta meta){
+    truncateLogIntern(count);
+    serializeMeta(meta);
+  }
+
+  private void truncateLogIntern(int count){
+    if(logSizeDeque.size() > count){
+      throw new IllegalArgumentException("truncate log count is bigger than total log count");
+    }
+
+    int size = 0;
+    for (int i = 0; i < count; i++) {
+      size += logSizeDeque.removeLast();
+    }
+    // write into disk
+    try {
+      logOutputStream.getChannel().truncate(logFile.length() - size);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public void removeFirst(int num) {
+    firstLogPosition += num;
+    for (int i = 0; i < num; i++) {
+      removedLogSize += logSizeDeque.removeFirst();
+    }
+
+    // do actual deletion
+    if (removedLogSize > maxRemovedLogSize) {
+      deleteRemovedLog();
+    }
+
+    // firstLogPosition changed
+    serializeMeta(meta);
```

Comment:
Are there any countermeasures against the situation that the logs are removed but the meta is not serialized?

Reply:
May be error log is the only method. Is there any suggestions?

Reply:
It would be easy if you use multiple files, in that case, you can serialize the metadata first and then remove the useless file. If the system crashes in between, during recovery you can know from the metadata that the log file is useless and still delete it.